Mathlib/Tactic/HaveLetLinter.lean

```diff
@@ -0,0 +1,114 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import Lean.Elab.Command
+import Lean.Linter.Util
+import Std.Data.List.Basic
+
+/-!
+#  The `have` vs `let` linter
+
+The `have` vs `let` linter flags uses of `have` to introduce a hypothesis whose Type is not `Prop`.
+
+TODO:
+* `replace` implies `have`: should it be ignored?
+* `haveI` may need to change to `let/letI`?
+* also do `let` vs `have`.
+-/
+
+open Lean Elab Command Meta
+
+namespace Mathlib.Linter
+
+/-- The `have` vs `let` linter emits a warning on `have`s introducing a hypothesis whose
+Type is not `Prop`. -/
+register_option linter.haveLet : Bool := {
+  defValue := true
+  descr := "enable the `have` vs `let` linter"
+}
+
+namespace haveLet
+
+/-- find the `have` syntax. -/
+partial
+def isHave? : Syntax → Bool
+  | .node _ ``Lean.Parser.Tactic.tacticHave_ _ => true
+  |_ => false
+
+/-- `SyntaxNodeKind`s that imply a `have` but should be ignored anyway. -/
+abbrev exclusions : HashSet SyntaxNodeKind := HashSet.empty
+  |>.insert ``Lean.Parser.Tactic.replace
+  |>.insert `Std.Tactic.classical!
+  |>.insert `Mathlib.Tactic.Tauto.tauto
+
+end haveLet
+
+end Mathlib.Linter
+
+namespace Mathlib.Linter.haveLet
+
+/-- a monadic version of `Lean.Elab.InfoTree.foldInfo`.
+Used to infer types inside a `CommandElabM`. -/
+def InfoTree.foldInfoM {α m} [Monad m] (f : ContextInfo → Info → α → m α) (init : α) :
+    InfoTree → m α :=
+  InfoTree.foldInfo (fun ctx i ma => do f ctx i (← ma)) (pure init)
+
+/-- given a `ContextInfo`, a `LocalContext` and an `Array` of `Expr`essions `es`,
+`areProp_toFormat` creates a `MetaM` context, and returns an array of pairs consisting of
+* a `Bool`ean, answering the question of whether the Type of `e` is a `Prop` or not, and
+* the pretty-printed `Format` of `e`
+for each `Expr`ession `e` in `es`.
+Concretely, `areProp_toFormat` runs `inferType` in `CommandElabM`.
+This is the kind of monadic lift that `nonPropHaves` uses to decide whether the Type of a `have`
+is in `Prop` or not.
+The output `Format` is just so that the linter displays a better message. -/
+def areProp_toFormat (ctx : ContextInfo) (lc : LocalContext) (es : Array Expr) :
+    CommandElabM (Array (Bool × Format)) := do
+  ctx.runMetaM lc do
+    es.mapM fun e => do
+      let typ ← inferType (← instantiateMVars e)
+      return (typ.isProp, ← ppExpr e)
+
+/-- returns the `have` syntax whose corresponding hypothesis does not have Type `Prop` and
+also a `Format`ted version of the corresponding Type. -/
+partial
+def nonPropHaves : InfoTree → CommandElabM (Array (Syntax × Format)) :=
+  InfoTree.foldInfoM (init := #[]) fun ctx info args => return args ++ (← do
+    let .ofTacticInfo i := info | return #[]
+    let stx := i.stx
+    let .original .. := stx.getHeadInfo | return #[]
```

Comment:
This is essentially the only extra change that I added.  Without it, the linter was still flagging, e.g., `replace`.  I think that the reason is that it was descending into the implementation of `replace`, rather than "staying on the surface".